src/gleam/list.gleam

```diff
@@ -248,6 +249,48 @@ pub fn rest(list: List(a)) -> Result(List(a), Nil) {
   }
 }
 
+fn group_updater(
+  f: fn(element) -> key,
+) -> fn(Map(key, List(element)), element) -> Map(key, List(element)) {
+  fn(groups, elem) {
+    case map.get(groups, f(elem)) {
+      Ok(existing) -> map.insert(groups, f(elem), [elem, ..existing])
+      Error(_) -> map.insert(groups, f(elem), [elem])
+    }
+  }
+}
+
+/// Takes a list and groups the values by a key
+/// which is built from a key function.
+///
+/// ## Examples
+///
+/// ```gleam
+/// > [Ok(3), Error("Wrong"), Ok(200), Ok(73)]
+///   |> group(with: fn(i) {
+///     case i {
+///       Ok(_) -> "Successful"
+///       Error(_) -> "Failed"
+///     }
+///   })
+///   |> map.to_list
+///
+/// [
+///   #("Failed", [Error("Wrong")]),
+///   #("Successful", [Ok(3), Ok(200), Ok(73)])
+/// ]
+///
+/// > group(from: [1,2,3,4,5], with: fn(i) {fn(i) { i - i / 3 * 3 }})
+/// |> map.to_list
+/// [#(0, [3]), #(1, [1, 4]), #(2, [2, 5])]
+/// ```
+///
+pub fn group(list: List(v), by key: fn(v) -> k) -> Map(k, List(v)) {
+  list
+  |> fold(map.new(), group_updater(key))
+  |> map.map_values(fn(_, group) { reverse(group) })
```

Comment:
How about we don't reverse the list order? Save some extra computation. The user can reverse it if they need that property.

Reply:
Yeah, I like the idea. We do not force a user to reverse the list when it is not needed. Will remove it.

Reply:
I think I found a solution that keeps the correct order without additional computation by using `list.fold_right` instead of `list.fold`.  The whole `map.values` part was superfluous from the beginning.

Reply:
We want to fold from the left as that is tail recursive. Folding from the right is not and uses an unbound amount of stack space, which may overflow.

Reply:
OK, then I will change it back to the "normal" fold and add a disclaimer to the docs that the function does not preserve the order.